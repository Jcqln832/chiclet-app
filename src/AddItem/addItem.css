.form--add {
    position: absolute;
    bottom: 0;
    left: 30px;
    right: 30px;
    margin-bottom: 30px;
    text-align: center;
    padding: 12px 0px;
}


.form--add__label {
    display: block;
    text-align: left;
    font-weight: bold;
    font-size: 14px;
    margin-bottom: 4px;
    padding-left: 6px;
}

.form--add__input {
    width: 100%;
    margin-bottom: 6px;
    font-size: 1em;
    height: 30px;
    box-sizing: border-box;
    padding: 6px;
    background-color: #ecebeb;
}

.form--add__submit {
    position: absolute;
    right: 0;
    top: 34px;
    height: 28px;
    border: 0px solid transparent;
    background: #b5b5b5;
    border-left: 1px solid;
    font-size: 14px;
}