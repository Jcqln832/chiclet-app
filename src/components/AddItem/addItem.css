.form--add {
    position: absolute;
    bottom: 0;
    left: 30px;
    right: 30px;
    margin-bottom: 30px;
    text-align: center;
    padding: 12px 0px;
}

.form--add__label {
    display: block;
    text-align: left;
    font-weight: bold;
    font-size: 14px;
    margin-bottom: 4px;
    padding-left: 6px;
}

.form--add__input {
    width: 100%;
    margin-bottom: 6px;
    font-size: 1em;
    height: 30px;
    box-sizing: border-box;
    padding: 6px;
    background-color: #fff;
}

.form--add__input:focus {
    outline-color: #ffc354;
    background-color: white;
    outline-width: 4px;
}

.form--add__submit {
    position: absolute;
    right: 0;
    top: 38px;
    height: 28px;
    border: 0px solid transparent;
    background: #ffc354;
    font-weight: bold;
    color: #011c33;
    border-left: 1px solid;
    font-size: 14px;
    transition: .2s;
}

.form--add__submit:hover {
    background: #ffc354;
}