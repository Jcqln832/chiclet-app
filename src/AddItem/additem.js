import React, { Component } from 'react';
import config from '../config';
import {withRouter} from 'react-router-dom';
import ValidationError from '../ValidationError';
import PropTypes from 'prop-types';
import './addItem.css'

class AddItem extends Component {
  constructor(props) {
    super(props);
    //capture form inputs here to make a controlled form
    this.state = {
      item: "",
      itemValid: false,
      formValid: false,
      validationMessages: {
        item: ''
      }
    }
  }

  itemChanged(item) {
    this.setState({item})
  }

  // Validate item not empty and not too long. Return a message if so
  validateItem(fieldValue) {
    fieldValue = fieldValue.trim();
    if (fieldValue.length === 0) {
        return 'item content is required'
    } else if 
     (fieldValue.length > 150) {
        return 'item content exceeds length limit (150)'
    } else {
        return "";
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    console.log("clicked!");
    // Get item from state
    const item = this.state.item;
    // Validate before fetch attempt
    this.validateForm(item);
  }

  validateForm(item) {
   const itemMessage =  this.validateItem(item);

    this.setState({
      validationMessages: {
        item: itemMessage,
      },
      itemValid: !itemMessage,
    }, this.formValid );
  }

  formValid() {
    this.setState({
      formValid: (this.state.itemValid)
    }, () => this.doFetch());
  }

  doFetch() {
    const url = `${config.API_ENDPOINT}/month/:monthId`;
    const item = this.state.item;
    const month= this.props.month;
    const index = this.props.monthIndex

    if(this.state.formValid) {
      console.log(this.state.formValid);
    const options = {
      method: 'POST',
      body: JSON.stringify({
          item: item,
          month: month,
          index: index
      }),
      headers: {
        "Content-Type": "application/json",
      }
    }
    console.log(options)

   this.props.addItem(
        {
            id: "",
            item: item,
            author: "signedinUser",
            month: month,
            index: Number(index)
          }
    )
}

    // put the data into STORE(db)
    // fetch(url, options)
    // .then(res => {
    //   if(!res.ok) {
    //     throw new Error('Something went wrong, please try again later.')
    //   }
    //   return res.json()
    // })
    // .then(data => {
    //   // clear for next entry
    //   this.setState({
    //     item: "",
    //     }, () => {this.props.addItem(  //add item from database into app state (incudling auto incremented item id#)
    //     {
    //         // id: data.id,
    //         // item: "data.item",
    //         // author: "data.author",
    //         // month: "data.month",
    //         // index: data.index
    //       },
    //     )}
    //   );
    //   console.log("I ran");
    // })
    // .catch(err => {
    //   this.setState({
    //     error: err.message
    //   });
    // });
//   }
}

  render () {
    const error = this.state.error ? <div className="error">{this.state.error}</div> : "";
    return (
        <section className='AddItem'>
            { error }
            <form className="form--add" onSubmit={e => this.handleSubmit(e)}>
            
                <div className="field">
                    <label class="form--add__label"htmlFor="item-input">Add an item</label>
                    <input class="form--add__input" type="text" name="name" id="item-input" aria-label="new month item" aria-required="true" onChange={e => this.itemChanged(e.target.value)}/>
                    {<ValidationError hasError={!this.state.nameValid} message={this.state.validationMessages.name}/>}
                </div>
                <button class="form--add__submit" type="submit">
                 + Save
                </button>
            </form>
        </section>
    )
 }
}

export default withRouter(AddItem);

AddItem.defaultProps = {
  history: {
    goBack: () => {}
  }
}

AddItem.propTypes = {
  addItem: PropTypes.func.isRequired
};