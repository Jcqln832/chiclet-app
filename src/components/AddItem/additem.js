import React, { Component } from 'react';
// import config from '../config';
import {withRouter} from 'react-router-dom';
import ValidationError from '../../ValidationError';
import apiContext from '../../apiContext';
import PropTypes from 'prop-types';
import './addItem.css'
import ItemApiService from '../../services/item-api-service';
// import MonthItems from '../months/monthitems';

class AddItem extends Component {
  static contextType = apiContext;
  constructor(props) {
    super(props);
    // capture form inputs here to make a controlled form
    this.state = {
      content: "",
      itemValid: false,
      formValid: false,
      error: "",
    }
  }

  itemChanged(content) {
    this.setState({content})
    console.log(this.state.content)
  }

  // Validate item not empty and not too long. Return a message if so.
  validateItem(fieldValue) {
    fieldValue = fieldValue.trim();
    if (fieldValue.length === 0) {
        return 'Item content is required'
    } else if 
     (fieldValue.length > 60) {
        return 'Item content exceeds length limit (60)'
    } else if 
      (this.props.monthItems.length === 6) {
        return 'Item limit reached. Remove an item to add another.'
      }
     else { 
        return false;
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    console.log("clicked!");
    // Get item from state
    const item = this.state.content;
    // Validate before fetch attempt
    this.validateForm(item)
  }

  validateForm(item) {
    const itemMessage =  this.validateItem(item);
    this.setState({
      error: itemMessage,
      itemValid: !itemMessage,
    }, this.formValid );
  }

  formValid() {
    this.setState({
      formValid: (this.state.itemValid)
    }, () => this.doFetch());
  }

   doFetch() {
    const content = this.state.content;
    const month= this.props.month;
    const index = Number(this.props.monthIndex)

    if(this.state.formValid) {
      // console.log(this.state.formValid);
      ItemApiService.postItem(content, index)
      .then(this.context.addItem)
      .then(this.setState({
        content: "",
        itemValid: false,
        formValid: false,
        error: "",
      }))
      .catch(this.context.setError)
  }
}

  render () {
    // const error = this.state.error ? <div className="error">{this.state.error}</div> : "";
    return (
        <section className='AddItem'>
          <form className="form--add" onSubmit={e => this.handleSubmit(e)}>
              <div className="field">
                <label className="form--add__label"htmlFor="item-input">Add an item</label>
                <input className="form--add__input" type="text" name="name" id="item-input" aria-label="new month item" aria-required="true" value={this.state.content} onChange={e => this.itemChanged(e.target.value)}/>
                {<ValidationError hasError={!this.state.itemValid} message={this.state.error}/>}
              </div>
              <button className="form--add__submit" type="submit">
                 + Save
              </button>
          </form>
        </section>
    )
  }
}

export default withRouter(AddItem);

AddItem.defaultProps = {
  history: {
    goBack: () => {}
  }
}