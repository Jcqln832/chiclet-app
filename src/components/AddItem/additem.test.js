import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import AddItem from './additem';
import ITEMS from '../../Utils/itemsList';
import testWrapper from '../../Utils/testHelpers';

describe('single month list items tests', () => {
    const div = document.createElement('div');
             
    it('renders without crashing', () => {
        ReactDOM.render(
            testWrapper(
                <AddItem />
            )
        , div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const tree = renderer
        .create( 
            testWrapper(
                <AddItem month={"January"} monthIndex={201901} addItem={()=>{}} monthItems={ITEMS.filter(item => item.index === 201901)}/>
            )
        )
          .toJSON();
        expect(tree).toMatchSnapshot();  
    });
});