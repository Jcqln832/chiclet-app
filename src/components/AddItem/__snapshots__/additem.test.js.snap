// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single month list items tests renders the UI as expected 1`] = `"component"`;
